pkg/bosh/manifest/dns.go

```diff
@@ -0,0 +1,316 @@
+package manifest
+
+import (
+	"context"
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+
+	appsv1 "k8s.io/api/apps/v1"
+
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"code.cloudfoundry.org/cf-operator/pkg/kube/util/mutate"
+
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"code.cloudfoundry.org/cf-operator/pkg/kube/util/names"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//DomainNameService abstraction
+type DomainNameService interface {
+	// FindServiceNames determines how a service should be named in accordance with the 'bosh-dns'-addon
+	FindServiceNames(instanceGroupName string) []string
+
+	// HeadlessServiceName constructs the headless service name for the instance group.
+	HeadlessServiceName(instanceGroupName string) string
+
+	// DNSSetting get the DNS settings for POD
+	DNSSetting() (corev1.DNSPolicy, *corev1.PodDNSConfig, error)
+
+	// Reconcile DNS stuff
+	Reconcile(ctx context.Context, namespace string, c client.Client, setOwner func(object metav1.Object) error) error
+}
+
+// Target of domain alias
+type Target struct {
+	Query         string `json:"query"`
+	InstanceGroup string `json:"instance_group" mapstructure:"instance_group"`
+	Deployment    string `json:"deployment"`
+	Network       string `json:"network"`
+	Domain        string `json:"domain"`
+}
+
+// Alias of domain alias
+type Alias struct {
+	Domain  string   `json:"domain"`
+	Targets []Target `json:"targets"`
+}
+
+// boshDomainNameService is used to emulate Bosh DNS
+type boshDomainNameService struct {
+	Aliases      []Alias
+	LocalDNSIP   string
+	ManifestName string
+}
+
+var _ DomainNameService = &boshDomainNameService{}
+
+// BoshDNSAddOnName name of bosh dns add on
+const BoshDNSAddOnName = "bosh-dns-aliases"
+
+// NewBoshDomainNameService create a new DomainNameService
+func NewBoshDomainNameService(addOn *AddOn, manifestName string) (DomainNameService, error) {
+	dns := boshDomainNameService{ManifestName: manifestName}
+	for _, job := range addOn.Jobs {
+		aliases := job.Properties.Properties["aliases"]
+		if aliases != nil {
+			var a = make([]Alias, 0)
+			err := mapstructure.Decode(aliases, &a)
+			if err != nil {
+				return nil, errors.Wrapf(err, "Loading aliases from manifest")
+			}
+			dns.Aliases = append(dns.Aliases, a...)
+		}
+	}
+	return &dns, nil
+}
+
+// FindServiceNames see interface
+func (dns *boshDomainNameService) FindServiceNames(instanceGroupName string) []string {
```

Comment:
Since we are doing DNS "aliasing", I don't think we should touch the services cf-operator currently creates. Is it possible to only create the extra service for the alias and keep the rest untouched?

Reply:
In my point of view, the new DNS resolution should be superfluous in the near future and should be replaced by the k8s DNS resolution. To work in this direction, it's essential to create the service with the name which corresponds to the DNS entry.
The alias with the manifest-name-prefix only exists to avoid breaking changes for existing BOSH deployments like [scf]((https://github.com/SUSE/scf/).
My vision would be to remove the aliases from the DNS resolution if all the DNS related stuff in the OPS files (e.g. [api.yaml](https://github.com/SUSE/scf/blob/98137edf63d15c2b0d49970172ac3b37d50bc096/deploy/helm/scf/assets/operations/instance_groups/api.yaml#L57) ) is cleaned up.

Reply:
I pushed a commit that changes how the rewrites work with the current services cf-operator creates. cf-operator prepends the deployment name because we can have more than one installation per namespace. Since the CoreDNS instance is per deployment, the generated rewrites will resolve the correct service. In the case of SCF, if no `.service.cf.internal` is used, the CoreDNS instance of the deployment will simply forward the request to the upstream server. I linked to this PR the draft PR I'm working on the SCF side.